Jess/webid help

jesscmoore opened the pull request:
# Pull Request Details

**REVIEW this PR with notepod branch jess/get_resources**

## What issue does this PR address

- Implements a clickable suggestion list in the Individual webid text input

- Link to associated issue: #295 
- See notes in comment https://github.com/anusii/solidpod/issues/295#issuecomment-3138435210 

## Checklist

I believe this is a non-breaking change. It adds an optional parameter comprising map of files and their ACLs to GrantPermissionUI. Apps such as NotePod which already fetch that data for ListNotes(), should pass the data to the sharing button call to GrantPermissionUI(). This avoids running a future to retrieve in when the Individual WebID button is clicked.

Tested on apps:
- [x] NotePod - in PR jess/get_resources - see screenshots in issue.
- [ ] Another app - please test on another app that calls GrantPermissionUI()

Complete the check-list below to ensure your branch is ready for PR.

Flutter Style Guide: https://survivor.togaware.com/gnulinux/flutter-style.html

- [x] Screenshots included in linked issue
- [x] Changes adhere to the team style and coding guideline
- [x] No confidential information
- [x] No duplicated content
- [x] No lint check errors related to your changes (`make prep` or `flutter analyze lib`)
- [ ] Pre-exisiting lint errors noted: [HERE]
- [x] Tested on at least one device
  - [ ] Android Phone
  - [ ] Android Emulator
  - [ ] Chrome on Android
  - [x] Chrome
  - [x] iOS
  - [ ] Linux
  - [ ] MacOS
  - [ ] Windows
- [x] Added 2 reviewers (or 1 for private repositories then they add another)

## Finalising

Once PR discussion is complete and 2 reviewers have approved:

- [ ] Merge dev into the branch
- [ ] Resolve any conflicts
- [ ] Add one line summary into CHANGELOG.md
- [ ] Bump appropriate version number in pubspec.yaml
- [ ] Push to git repository and review
- [ ] Merge PR into dev

jesscmoore committed:
replace literal with constant

jesscmoore committed:
replace literal with constant

jesscmoore committed:
retrieve list of unique recipient webIds

jesscmoore committed:
implemented webid suggestions drop down

jesscmoore committed:
use boxedSuggestionList() that wraps suggestion list in box

jesscmoore committed:
tidied steps and passes data from calling app to avoid repeated fetching data for recipient list

jesscmoore committed:
Merge branch 'dev' into jess/webid_help

anushkavidanage commented:
Thanks Jess for these updates. Tested this with Notepod and this looks and worked great. I have a couple of observations/suggestions. For reference, see the screenshot below.

1. Currently the input dialog takes up the whole screen which might look bit strange for the user. Can we make the dislog dimension smaller?
2. I don't know if you can see, but I see some grey lines in the background of the dialog. They are extremely faded, but I see them. I wonder what is causing that.
3. The scrollbar is in the middle of the dialog. Is it possible to move that to the right end?
4. When I type a name without entering the `https:` part, the suggestions disappear. For instance, in the below instance, when I type `will` the suggestion for `will-smith` web id disappears as well.  We probably want to filter the suggestions by checking the entire web ID with the typed content?

<img width="1920" height="1050" alt="webid-input" src="https://github.com/user-attachments/assets/484c1d46-b9dd-4c84-afe5-f1b65f5c6b56" />

I will do some more testing with another app as well.

jesscmoore commented:
Thanks Anushka for testing the changes. I'll get onto these observations as soon as I can and make adjustments.
> 
> 1. Currently the input dialog takes up the whole screen which might look bit strange for the user. Can we make the dislog dimension smaller?
Yes, currently it is set as fraction of the media width, hence it looks wider on your different platform. I'll fix to provide settings that work on different platforms. Were you building it to Windows platform?

> 2. I don't know if you can see, but I see some grey lines in the background of the dialog. They are extremely faded, but I see them. I wonder what is causing that.
I don't know if I've noticed that in the webid input dialog. But I do 2 shades of background in the loading indicator dialog which needs fixing. I can have a look at both dialogs to see what's causing these background issues.

> 3. The scrollbar is in the middle of the dialog. Is it possible to move that to the right end?
Yes, will fix that. It wasn't apparent on iOS/chrome tests I did.

> 4. When I type a name without entering the `https:` part, the suggestions disappear. For instance, in the below instance, when I type `will` the suggestion for `will-smith` web id disappears as well.  We probably want to filter the suggestions by checking the entire web ID with the typed content?
I'll have to test further to see what needs changing. Currently the suggestions show any that contain the search parameter, so I would have expected a match of will to webids with will in them.

> I will do some more testing with another app as well.
Great, thank you Anushka

jesscmoore committed:
use listview.buiilder with elevated card instead of listview.separated with divider

jesscmoore committed:
removed padding on sides of suggestions

jesscmoore committed:
added instruction

jesscmoore committed:
moved constants to layout class and use auto dialog width sizing

jesscmoore committed:
uses full width on iphone/android, else uses specified width

jesscmoore commented:
@anushkavidanage PR updated for you - see below

> > 1. Currently the input dialog takes up the whole screen which might look bit strange for the user. Can we make the dislog dimension smaller?
Done, uses set narrower width, unless on iOS/android in which case full width is used.

> > 2. I don't know if you can see, but I see some grey lines in the background of the dialog. They are extremely faded, but I see them. I wonder what is causing that.
I was using Listview.separated to add a divider between suggestions. Switched to standard Listview and using card to indicate that suggestions are clickable and remove divider lines.

> > 3. The scrollbar is in the middle of the dialog. Is it possible to move that to the right end?
Fixed. Suggestion widget now uses full width of dialog such that scroll bar is right aligned correctly now.

> > 4. When I type a name without entering the `https:` part, the suggestions disappear. For instance, in the below instance, when I type `will` the suggestion for `will-smith` web id disappears as well.  We probably want to filter the suggestions by checking the entire web ID with the typed content?
The filtering is on a contains basis. such that if you type will, webids containing will, will-smith, will-etc will all show. Can you check again.

anushkavidanage commented on lib/src/widgets/ind_webid_input_dialog.dart:
```diff
@@ -96,38 +118,73 @@ class _IndWebIdTextInputState extends State<IndWebIdTextInput> {
+      return;
+    }
+    suggestionList =
+        webIdList.where((e) => e.contains(value.toLowerCase())).toList();
```
I think we should also convert `e` to lowercase as well. For instance I have a web ID which contains the string `Gerry-Tonga`. This does not filter out even when I type `Gerry` because we are only converting the value to lowercase.

jesscmoore replied:
Thanks @anushkavidanage for the review and catching this. I've pushed this improvement. 

Currently unable to login to test on iPad as below - is this a known new issue?

```
Restarted application in 846ms.
flutter: AuthDataManager => _getTokenResponse() failed: OpenIdException(invalid_grant): grant request is invalid
flutter: AuthDataManager => getAccessToken() failed
flutter: AuthDataManager => _getTokenResponse() failed: OpenIdException(invalid_grant): grant request is invalid
flutter: AuthDataManager => getAccessToken() failed
flutter: solidAuthenticate() => solid_auth.authenticate(https://pods.solidcommunity.au)
flutter: Solid Authenticate Failed: SocketException: Failed to create server socket (OS Error: Address already in use, errno = 48), address = 0.0.0.0, port = 4400
```

anushkavidanage replied:
Hmm, haven't seen this issue before. Are you running multiple apps at once?

jesscmoore replied:
You're quite right. Works fine without multiple simulators open.

anushkavidanage reviewed (approved):
Everything looks good Jess. Thanks for making the changes. I added only one minor comment.

jesscmoore committed:
fixed lower case matching

gjwgit reviewed (approved):
Very nice. Thanks @jesscmoore Tested with notepod.

gjwgit commented:
Published.
